### PR TITLE
Perf doc v5

### PR DESCRIPTION
**Pull request by pevma**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

- https://redmine.openinfosecfoundation.org/issues/2839

Describe changes:
Addresses changes/reviews in https://github.com/OISF/suricata/pull/4182 , https://github.com/OISF/suricata/pull/4185  and  doc build errs https://travis-ci.org/OISF/suricata/jobs/585984036

- doc: Update high performance config doc
- doc: Update tuning considerations doc

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

File: doc/userguide/performance/high-performance-config.rst [hunk 1]
```diff
@@ -1,19 +1,379 @@
...
 ==============================
 
-If you have enough RAM, consider the following options in suricata.yaml to off-load as much work from the CPU's as possible:
+NIC
+---
+
+One of the major dependencies for Suricata's performance is the Network 
+Interface Card. There are many vendors and possibilities. Some NICs have and 
+require their own specific instructions and tools of how to set up the NIC. 
+That way ensuring the greatest benefit when running Suricata. Vendors like
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: Suggest `This ensures the greatest benefit when running Suricata`

File: doc/userguide/performance/high-performance-config.rst [hunk 2]
```diff
@@ -1,19 +1,379 @@
...
+available per port/interface) the worker threads can be set up accordingly. 
+Usually the available queues can be seen by running:
+
+::
+
+ /usr/local/sbin/ethtool -l eth1
+
+Some NICs - generally lower end 1Gbps - do not support symmetric hashing see 
+:doc:`packet-capture`. On those systems due to considerations for out of order 
+packets the following set up with af-packet is suggested (the example below
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: nit s/set up/setup

File: doc/userguide/performance/high-performance-config.rst [hunk 3]
```diff
@@ -1,19 +1,379 @@
...
+ /usr/local/sbin/ethtool -A eth1 rx off 
+ /usr/local/sbin/ethtool -C eth1 adaptive-rx off adaptive-tx off rx-usecs 125
+ /usr/local/sbin/ethtool -G eth1 rx 1024
+
+The commands above can be reviewed in detail in the help or manpages of the 
+``ethtool``. In brief the sequence makes sure the NIC is reset, the number of 
+RSS queues is set to 16, load balancing is enabled for the NIC, a low entropy 
+toepiltz key is inserted to allow for symmetric hashing, receive offloading is 
+disabled, the adaptive control is disabled for lowest possible latency and 
+last but not least, the ring rx descriptor size.
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: Suggest: `the ring rx descriptor size is set to 1024`.

File: doc/userguide/performance/high-performance-config.rst [hunk 4]
```diff
@@ -1,19 +1,379 @@
...
+
+ for proto in tcp4 udp4 tcp6 udp6; do
+    /usr/local/sbin/ethtool -N eth1 rx-flow-hash $proto sdfn
+ done
+
+In some cases:
+
+::
+
+ /usr/local/sbin/ethtool -N eth1 rx-flow-hash $proto sd
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: Should you mention the difference between `sdfn` and `sd`?

File: doc/userguide/performance/high-performance-config.rst [hunk 5]
```diff
@@ -1,19 +1,379 @@
...
+
+CPU affinity and NUMA
+---------------------
+
+Intel based systems
+~~~~~~~~~~~~~~~~~~~
+
+If the system has more then one NUMA node there are some more possibilities. 
+In those cases it is generally recommended to use as many worker threads as 
+cpu cores available/possible - from the same NUMA node. The example below uses
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: nit: s/uses/uses a/

File: doc/userguide/performance/high-performance-config.rst [hunk 6]
```diff
@@ -1,19 +1,379 @@
...
+    /usr/local/sbin/ethtool -X eth1 hkey 6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A equal 36
+    /usr/local/sbin/ethtool -A eth1 rx off tx off
+    /usr/local/sbin/ethtool -C eth1 adaptive-rx off adaptive-tx off rx-usecs 125
+    /usr/local/sbin/ethtool -G eth1 rx 1024
+    for proto in tcp4 udp4 tcp6 udp6; do
+        echo "/usr/local/sbin/ethtool -N eth1 rx-flow-hash $proto sdfn"
+        /usr/local/sbin/ethtool -N eth1 rx-flow-hash $proto sdfn
+    done
+
+In the example above the ``set_irq_affinity`` script is used form the NIC
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: nit: s/form/from/

File: doc/userguide/performance/high-performance-config.rst [hunk 7]
```diff
@@ -1,19 +1,379 @@
...
+    /usr/local/sbin/ethtool -A eth1 rx off tx off
+    /usr/local/sbin/ethtool -C eth1 adaptive-rx off adaptive-tx off rx-usecs 125
+    /usr/local/sbin/ethtool -G eth1 rx 1024
+    for proto in tcp4 udp4 tcp6 udp6; do
+        echo "/usr/local/sbin/ethtool -N eth1 rx-flow-hash $proto sdfn"
+        /usr/local/sbin/ethtool -N eth1 rx-flow-hash $proto sdfn
+    done
+
+In the example above the ``set_irq_affinity`` script is used form the NIC 
+drivers sources.
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: nit: s/drivers/driver's/

File: doc/userguide/performance/high-performance-config.rst [hunk 8]
```diff
@@ -1,19 +1,379 @@
...
+in total per CPUs NUMA 1 range - 18-35,54-71. That part is done via the  
+``worker-cpu-set`` affinity settings. ``ring-size`` and ``block-size`` in the 
+config section  above are decent default values to start with. Those can be 
+better adjusted if needed as explained in :doc:`tuning-considerations`.
+    
+AMD based systems
+~~~~~~~~~~~~~~~~~
+
+Another example can be using an AMD based system where the architecture and 
+design of the system itself plus the NUMA nodes interaction is different as
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: nit: s/nodes/node's/

File: doc/userguide/performance/high-performance-config.rst [hunk 9]
```diff
@@ -1,19 +1,379 @@
...
+    mmap-locked: yes
+    tpacket-v3: yes
+    ring-size: 100000
+    block-size: 1048576
+
+
+In the example above there are 15 RSS queues pinned to cores 1-7,64-71 on NUMA 
+node 0 and 40 worker threads using other CPUs on different NUMA nodes. The 
+reason why CPU 0 is skipped in this set up is as in Linux systems it is very 
+common CPU 0 to be used by default by many tools/services. The NIC itself in
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: nit: s/common CPU/common for CPU/

File: doc/userguide/performance/high-performance-config.rst [hunk 10]
```diff
@@ -1,19 +1,379 @@
...
+    ring-size: 100000
+    block-size: 1048576
+
+
+In the example above there are 15 RSS queues pinned to cores 1-7,64-71 on NUMA 
+node 0 and 40 worker threads using other CPUs on different NUMA nodes. The 
+reason why CPU 0 is skipped in this set up is as in Linux systems it is very 
+common CPU 0 to be used by default by many tools/services. The NIC itself in 
+this config is positioned on NUMA 0 so starting with 15 RSS queues on that 
+NUMA node and keeping those off for otehr tools in the system could offer the
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: nit:s/otehr/other/

File: doc/userguide/performance/high-performance-config.rst [hunk 11]
```diff
@@ -1,19 +1,379 @@
...
+.. note:: Performance and optimization of the whole system can be affected upon regular NIC driver and pkg/kernel upgrades so it should be monitored regularly and tested out in QA/test environments first. As a general suggestion it is always recommended to run the latest stable firmware and drivers as  instructed and provided by the particular NIC vendor. 
 
-Be advised, however, that this may require lots of RAM for even modestly sized rule sets.  Also be aware that having additional CPU's available provides a greater performance boost than having more RAM available.  That is, it would be better to spend money on CPU's instead of RAM when configuring a system.
+Other considerations
+~~~~~~~~~~~~~~~~~~~~
 
-It may also lead to significantly longer rule loading times.
+Another advanced option to consider is the ``isolcpus`` kernel boot parameter 
+is a way of allowing CPU cores to be isolated for use of general system 
+processes. That way ensuring total dedication of those CPUs/ranges for the
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: nit: s/ensuring/ensures/

File: doc/userguide/performance/high-performance-config.rst [hunk 12]
```diff
@@ -1,19 +1,379 @@
...
+~~~~~~~~~~~~~~~~~~~~
 
-It may also lead to significantly longer rule loading times.
+Another advanced option to consider is the ``isolcpus`` kernel boot parameter 
+is a way of allowing CPU cores to be isolated for use of general system 
+processes. That way ensuring total dedication of those CPUs/ranges for the 
+Suricata process only.
 
+``stream.wrong_thread`` / ``tcp.pkt_on_wrong_thread`` are counters available
+in stats.log or eve.json as ``event_type: stats`` that indicate issues with
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: nit: delimit `stats.log` and `eve.json` with double ``

File: doc/userguide/performance/high-performance-config.rst [hunk 13]
```diff
@@ -1,19 +1,379 @@
...
+the load balancing. This could be traffic/NICs settings related as well. In
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: nit: s/This/There/

**Comment by pevma** [comment]
Comments addressed in https://github.com/OISF/suricata/pull/4264

**Pull request closed without merging**